### PR TITLE
Message Channel EIP

### LINKED ISSUE #143: Message Channel (EIP)

### PR DESCRIPTION
**Pull request by iluwatar**
JDP core team, please review. This is the first implemented EIP. I've used Apache Camel for the implementation. 

Resolves #143

### CODE REVIEW

**Comment by mafagafogigante** [comment]
Any way to improve the distribution of the elements of the image?

**Comment by iluwatar** [comment]
Hand adjusted the class diagram. Should be more readable now.

File: message-channel/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+/target/
+/target/
```
Comment (mafagafogigante) [inline message-channel/.gitignore]: @iluwatar Why duplicated?

File: message-channel/index.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+folder: message-channel
+permalink: /patterns/message-channel/
+categories: Integration
+tags: Java
+---
+
+**Intent:** When two applications communicate using a messaging system they do it by using logical addresses
+of the system, so called Message Channels.
+
+![alt text](./etc/message-channel.png "Message Channel")
```
Comment (mafagafogigante) [inline message-channel/index.md]: Doesn't ``Message Channel`` produces the same result? It is clearer, shorter and more readable.

File: message-channel/index.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+**Applicability:** Use the Message Channel pattern when
```
Comment (mafagafogigante) [inline message-channel/index.md]: Should we really **ensure** that all DPs have an applicability field? In this case it is redundant (Intent already said this).

File: message-channel/pom.xml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+		<artifactId>java-design-patterns</artifactId>
+		<version>1.5.0</version>
+	</parent>
+	<artifactId>message-channel</artifactId>
+	<dependencies>
+		<dependency>
+			<groupId>org.apache.camel</groupId>
+			<artifactId>camel-core</artifactId>
+		</dependency>
+			<dependency>
```
Comment (mafagafogigante) [inline message-channel/pom.xml]: Why this dependency is more indentated?

**Comment by mafagafogigante** [comment]
The diagram is better now. Did some code review.

**Comment by mafagafogigante** [comment]
Will this commits be squashed together before merging, right?

File: message-channel/src/main/java/com/iluwatar/message/channel/App.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import org.apache.camel.builder.RouteBuilder;
+import org.apache.camel.impl.DefaultCamelContext;
+
+/**
+ * 
+ * When two applications communicate with each other using a messaging system
+ * they first need to establish a communication channel that will carry the
+ * data. Message Channels are logical addresses in the messaging system.
+ * <p>
+ * The sending application doesn't necessarily know what particular application
```
Comment (npathai) [inline message-channel/src/main/java/com/iluwatar/message/channel/App.java]: Another way of saying that would be that "`Message Channel` decouples Message producers and Consumers".

File: message-channel/src/main/java/com/iluwatar/message/channel/App.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	 * @param args command line args
+	 * @throws Exception
+	 */
+	public static void main(String[] args) throws Exception {
+		CamelContext context = new DefaultCamelContext();
+		
+		context.addRoutes(new RouteBuilder() {
+
+			@Override
+			public void configure() throws Exception {
```
Comment (npathai) [inline message-channel/src/main/java/com/iluwatar/message/channel/App.java]: Can we have one more route, from `stream:in` to `direct:greetings`? This places emphasis on the point that the two endpoints don't need to know about each other.

Comment (npathai) [inline message-channel/src/main/java/com/iluwatar/message/channel/App.java]: This EIP being so trivial I scratched my head thinking of some particular use case but couldn't think of one without complicating the example. Do you have anything specific in mind?

File: message-channel/src/main/java/com/iluwatar/message/channel/App.java (hunk 5 above)
Comment (mafagafogigante) [inline message-channel/src/main/java/com/iluwatar/message/channel/App.java]: Yes. But if producers is lowercase, consumers should also be. Not a problem in the comment, but wouldn't be right on the code.

File: message-channel/src/main/java/com/iluwatar/message/channel/App.java (hunk 6 above)
Comment (mafagafogigante) [inline message-channel/src/main/java/com/iluwatar/message/channel/App.java]: I would also like some example use cases. A real world one would be awesome.

File: message-channel/src/main/java/com/iluwatar/message/channel/App.java (hunk 5 above)
Comment (npathai) [inline message-channel/src/main/java/com/iluwatar/message/channel/App.java]: :+1: a typo

File: message-channel/index.md (hunk 2 above)
Comment (iluwatar) [inline message-channel/index.md]: I think the implementation is the same across all the patterns and we have no agreed convention on this.

File: message-channel/index.md (hunk 3 above)
Comment (iluwatar) [inline message-channel/index.md]: I wouldn't leave it out. We should strive to have content for the fields mentioned in the [pattern template](https://github.com/iluwatar/java-design-patterns/wiki/2.-Pattern-template).

File: message-channel/src/main/java/com/iluwatar/message/channel/App.java (hunk 5 above)
Comment (iluwatar) [inline message-channel/src/main/java/com/iluwatar/message/channel/App.java]: This is done.

File: message-channel/src/main/java/com/iluwatar/message/channel/App.java (hunk 6 above)
Comment (iluwatar) [inline message-channel/src/main/java/com/iluwatar/message/channel/App.java]: I added the second message channel and a real world example.

**Comment by mafagafogigante** [comment]
Will you squash this before merging to keep the story clean?

**Comment by mafagafogigante** [comment]
Also, you can do `git pull --rebase ...` to avoid that useless merge commit.

**Comment by iluwatar** [comment]
See [Reviewing pull requests](https://github.com/iluwatar/java-design-patterns/wiki/8.-Reviewing-pull-requests).

**Comment by mafagafogigante** [comment]
So screw VCS history readability? Fantastic.

**Comment by iluwatar** [comment]
@mafagafogigante At the moment we have not agreed on requirement to squash everything before merge. We can discuss this on Gitter.

**Pull request merged**